// typia/lib/programmers/json/JsonApplicationProgrammer.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonApplicationProgrammer = void 0;
var TransformerError_1 = require("../../transformers/TransformerError");
var AtomicPredicator_1 = require("../helpers/AtomicPredicator");
var application_schema_1 = require("../internal/application_schema");
var JsonApplicationProgrammer;
(function (JsonApplicationProgrammer) {
    /**
     * @internal
     */
    var IOptions;
    (function (IOptions) {
        IOptions.complement = function (options) {
            var _a, _b;
            return ({
                purpose: (_a = options === null || options === void 0 ? void 0 : options.purpose) !== null && _a !== void 0 ? _a : "swagger",
                surplus: (_b = options === null || options === void 0 ? void 0 : options.surplus) !== null && _b !== void 0 ? _b : false,
            });
        };
    })(IOptions = JsonApplicationProgrammer.IOptions || (JsonApplicationProgrammer.IOptions = {}));
    JsonApplicationProgrammer.write = function (options) {
        return function (metadatas) {
            var fullOptions = IOptions.complement(options);
            var components = {
                schemas: {},
            };
            var generator = (0, application_schema_1.application_schema)(fullOptions)(true)(components);
            return __assign({ schemas: metadatas.map(function (meta, i) {
                    var schema = generator(meta)({});
                    if (schema === null)
                        throw new TransformerError_1.TransformerError({
                            code: "typia.json.application",
                            message: "invalid type on argument - (".concat(meta.getName(), ", ").concat(i, ")"),
                        });
                    return schema;
                }), components: components }, fullOptions);
        };
    };
    JsonApplicationProgrammer.validate = function (meta) {
        var e_1, _a;
        var output = [];
        if (meta.atomics.some(function (a) { return a.type === "bigint"; }) ||
            meta.constants.some(function (c) { return c.type === "bigint"; }))
            output.push("JSON schema does not support bigint type.");
        if (meta.tuples.some(function (t) {
            return t.type.elements.some(function (e) { return e.isRequired() === false; });
        }) ||
            meta.arrays.some(function (a) { return a.type.value.isRequired() === false; }))
            output.push("JSON schema does not support undefined type in array.");
        if (meta.maps.length)
            output.push("JSON schema does not support Map type.");
        if (meta.sets.length)
            output.push("JSON schema does not support Set type.");
        try {
            for (var _b = __values(meta.natives), _c = _b.next(); !_c.done; _c = _b.next()) {
                var native = _c.value;
                if (AtomicPredicator_1.AtomicPredicator.native(native) === false &&
                    native !== "Date" &&
                    native !== "Blob" &&
                    native !== "File")
                    output.push("JSON schema does not support ".concat(native, " type."));
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return output;
    };
})(JsonApplicationProgrammer || (exports.JsonApplicationProgrammer = JsonApplicationProgrammer = {}));
